File: src/parsing/functions/to_float.ts
import Function from "./function";
import { FunctionDef } from "./function_metadata";

@FunctionDef({
    description: "Converts a value to a floating point number",
    category: "scalar",
    parameters: [
        {
            name: "value",
            description: "The value to convert to a float",
            type: "any",
        },
    ],
    output: {
        description: "The floating point representation of the value",
        type: "number",
        example: "3.14",
    },
    examples: ['RETURN toFloat("3.14")', "RETURN toFloat(42)", "RETURN toFloat(true)"],
})
class ToFloat extends Function {
    constructor() {
        super("tofloat");
        this._expectedParameterCount = 1;
    }

    public value(): any {
        const val = this.getChildren()[0].value();
        if (val === null || val === undefined) {
            return null;
        }
        if (typeof val === "boolean") {
            return val ? 1.0 : 0.0;
        }
        if (typeof val === "number") {
            return val;
        }
        if (typeof val === "string") {
            const trimmed = val.trim();
            const parsed = Number(trimmed);
            if (isNaN(parsed)) {
                throw new Error(`Cannot convert string "${val}" to float`);
            }
            return parsed;
        }
        throw new Error("toFloat() expects a number, string, or boolean");
    }
}

export default ToFloat;

File: flowquery-py/src/parsing/functions/to_float.py
"""ToFloat function."""

from typing import Any

from .function import Function
from .function_metadata import FunctionDef


@FunctionDef({
    "description": "Converts a value to a floating point number",
    "category": "scalar",
    "parameters": [
        {"name": "value", "description": "The value to convert to a float", "type": "any"}
    ],
    "output": {"description": "The floating point representation of the value", "type": "number", "example": 3.14},
    "examples": [
        'RETURN toFloat("3.14")',
        "RETURN toFloat(42)",
        "RETURN toFloat(true)"
    ]
})
class ToFloat(Function):
    """ToFloat function.

    Converts a value to a floating point number.
    """

    def __init__(self) -> None:
        super().__init__("tofloat")
        self._expected_parameter_count = 1

    def value(self) -> Any:
        val = self.get_children()[0].value()
        if val is None:
            return None
        if isinstance(val, bool):
            return 1.0 if val else 0.0
        if isinstance(val, (int, float)):
            return float(val)
        if isinstance(val, str):
            trimmed = val.strip()
            try:
                return float(trimmed)
            except (ValueError, OverflowError):
                raise ValueError(f'Cannot convert string "{val}" to float')
        raise ValueError("toFloat() expects a number, string, or boolean")
